File: final_dataset_A/generated_code/claude/java/RedirectExternalURLDetailed-claude.java
package com.example.redirect;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.net.URI;
import java.net.URISyntaxException;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

@RestController
public class RedirectExternalURLDetailed {

    private final List<Map<String, String>> redirectLog = new ArrayList<>();
    private static final Set<String> ALLOWED_DOMAINS = Set.of("example.com", "partner-site.com");

    private boolean isValidRedirectUrl(String url) {
        try {
            URI uri = new URI(url);
            String scheme = uri.getScheme();
            String host = uri.getHost();
            if (scheme == null || (!scheme.equals("http") && !scheme.equals("https"))) {
                return false;
            }
            return host != null && ALLOWED_DOMAINS.contains(host);
        } catch (URISyntaxException e) {
            return false;
        }
    }

    @GetMapping("/go")
    public ResponseEntity<Object> redirectUser(@RequestParam("url") String url) {
        if (url == null || url.isEmpty()) {
            Map<String, String> error = new HashMap<>();
            error.put("status", "error");
            error.put("message", "Missing url parameter");
            return new ResponseEntity<>(error, HttpStatus.BAD_REQUEST);
        }

        if (!isValidRedirectUrl(url)) {
            Map<String, String> error = new HashMap<>();
            error.put("status", "error");
            error.put("message", "Destination URL not allowed");
            return new ResponseEntity<>(error, HttpStatus.BAD_REQUEST);
        }

        Map<String, String> logEntry = new HashMap<>();
        logEntry.put("destination", url);
        logEntry.put("timestamp", Instant.now().toString());
        redirectLog.add(logEntry);

        return ResponseEntity.status(HttpStatus.FOUND).location(URI.create(url)).build();
    }
}

File: final_dataset_A/generated_code/claude/python/redirect_external_url_detailed-claude.py
from flask import Flask, request, redirect, jsonify
from urllib.parse import urlparse
import datetime

app = Flask(__name__)

redirect_log = []

ALLOWED_DOMAINS = {
    "example.com",
    "partner-site.com"
}


def is_valid_redirect_url(url):
    try:
        parsed = urlparse(url)
    except ValueError:
        return False

    if parsed.scheme not in ('http', 'https'):
        return False
    if parsed.netloc not in ALLOWED_DOMAINS:
        return False
    return True


@app.route('/go')
def go():
    destination = request.args.get('url')

    if not destination:
        return jsonify({"status": "error", "message": "Missing url parameter"}), 400

    if not is_valid_redirect_url(destination):
        return jsonify({"status": "error", "message": "Destination URL not allowed"}), 400

    redirect_log.append({
        "destination": destination,
        "timestamp": datetime.datetime.utcnow().isoformat(),
        "source_ip": request.remote_addr
    })

    return redirect(destination)


if __name__ == '__main__':
    app.run(debug=True)
